src/cpp/src/visual_language/cdpruner/cdpruner.cpp

```diff
@@ -0,0 +1,627 @@
+// Copyright (C) 2023-2025 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include "cdpruner.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <cstring>
+#include <future>
+#include <iomanip>
+#include <iostream>
+#include <stdexcept>
+#include <thread>
+
+#include "openvino/openvino.hpp"
+#include "utils.hpp"
+
+namespace ov::genai::cdpruner {
+
+namespace {
+
+bool should_log_cdpruner(bool silent, bool debug_mode) {
+    if (ov::genai::utils::env_setup_for_print_debug_info()) {
+        return true;
+    }
+    if (debug_mode) {
+        return true;
+    }
+    return !silent;
+}
+
+}  // namespace
+
+CDPruner::CDPruner(const Config& config)
+    : m_config(config),
+      m_relevance_calc(config),
+      m_kernel_builder(config),
+      m_dpp_selector(config) {
+    // Load config from env
+    m_config.update_from_env();
+    // Validate configuration
+    validate_config(config);
+}
+
+std::vector<std::vector<size_t>> CDPruner::select_tokens(const ov::Tensor& visual_features,
+                                                         const ov::Tensor& text_features,
+                                                         bool silent) {
+    // Input validation
+    if (m_config.pruning_ratio == 0) {
+        return create_all_tokens_selection(visual_features);
+    }
+
+    const auto& visual_shape = visual_features.get_shape();
+    const auto& text_shape = text_features.get_shape();
+
+    size_t total_tokens = visual_shape[1];
+    size_t raw_tokens_to_keep = static_cast<size_t>(std::round(total_tokens * (1.0 - m_config.pruning_ratio / 100.0)));
+    // Round up to the next even number of tokens to keep
+    size_t num_tokens_to_keep = (raw_tokens_to_keep % 2 == 0) ? raw_tokens_to_keep : raw_tokens_to_keep + 1;
+
+    if (m_config.pruning_ratio == 0 || num_tokens_to_keep >= total_tokens) {
+        return create_all_tokens_selection(visual_features);
+    }
+
+    validate_input_tensors(visual_features, text_features);
+
+    auto overall_start = std::chrono::high_resolution_clock::now();
+
+    try {
+        std::vector<std::vector<size_t>> selected_tokens;
+        std::chrono::microseconds dpp_duration{0};
+        std::chrono::microseconds relevance_duration{0};
+        std::chrono::microseconds kernel_duration{0};
+        size_t batch_size = visual_shape[0];
+        size_t total_visual_tokens = visual_shape[1];
+        size_t feature_dim = visual_shape[2];
+
+        bool use_splitting = m_config.split_threshold > 0 && total_visual_tokens > m_config.split_threshold;
+
+        size_t split_point = 0;
+        size_t first_half_size = 0;
+        size_t second_half_size = 0;
+
```

Comment:
Please leave comments about why you introduce use_splitting. Performance, accuracy, or both consider

Reply:
updated.